Dart/src/com/jetbrains/lang/dart/ide/refactoring/ServerRefactoring.java

```diff
@@ -0,0 +1,216 @@
+  private final int offset;
+  private final int length;
+
+  private final Set<Integer> pendingRequestIds = Sets.newHashSet();
+  private RefactoringStatus serverErrorStatus;
+  private RefactoringStatus initialStatus;
+  private RefactoringStatus optionsStatus;
+  private RefactoringStatus finalStatus;
+  private SourceChange change;
+
+  private int lastId = 0;
+  private ServerRefactoringListener listener;
+
+  public ServerRefactoring(String kind, String name, String file, int offset, int length) {
+    this.kind = kind;
+    this.name = name;
+    this.file = file;
+    this.offset = offset;
+    this.length = length;
+  }
+
+  public RefactoringStatus checkFinalConditions() {
+    ProgressManager.getInstance().run(new Task.Modal(null, "Checking initial conditions", true) {
+      @Override
+      public void run(@NotNull ProgressIndicator indicator) {
+        setOptions(false, indicator);
+      }
+    });
+    if (serverErrorStatus != null) {
+      return serverErrorStatus;
+    }
+    return finalStatus;
+  }
+
+  public RefactoringStatus checkInitialConditions() {
+    ProgressManager.getInstance().run(new Task.Modal(null, "Checking initial conditions", true) {
+      @Override
+      public void run(@NotNull ProgressIndicator indicator) {
+        setOptions(true, indicator);
+      }
+    });
+    if (serverErrorStatus != null) {
+      return serverErrorStatus;
+    }
+    return initialStatus;
+  }
+
+  public SourceChange getChange() {
+    return change;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  /**
+   * Returns this {@link RefactoringOptions} subclass instance.
+   */
+  protected abstract RefactoringOptions getOptions();
+
+  /**
+   * Sets the received {@link RefactoringFeedback}.
+   */
+  protected abstract void setFeedback(RefactoringFeedback feedback);
+
+  public void setListener(ServerRefactoringListener listener) {
+    this.listener = listener;
+  }
+
+  protected void setOptions(boolean validateOnly, ProgressIndicator indicator) {
+    // add a new pending request ID
+    final int id;
+    synchronized (pendingRequestIds) {
+      id = ++lastId;
+      pendingRequestIds.add(id);
+    }
+    // do request
+    serverErrorStatus = null;
+    final CountDownLatch latch = new CountDownLatch(1);
+    RefactoringOptions options = getOptions();
+    final boolean success = DartAnalysisServerService.getInstance()
+      .edit_getRefactoring(kind, file, offset, length, validateOnly, options, new GetRefactoringConsumer() {
+        @Override
+        public void computedRefactorings(List<RefactoringProblem> initialProblems,
+                                         List<RefactoringProblem> optionsProblems,
+                                         List<RefactoringProblem> finalProblems,
+                                         RefactoringFeedback feedback,
+                                         SourceChange _change,
+                                         List<String> potentialEdits) {
```

Comment:
If I add `Thread.sleep(1000);` here to emulate slow response from server DartRenameDialog disables Refactor button forever without any error message after typing a new name. @scheglov can you please take a look at it?

FYI: Registry key now enables server driven resolution/find usages/rename in one shot, so you do not have to tweak plugin.xml to test it.